File: android/app/src/main/kotlin/com/torvm/android/packet/PacketBuilder.kt
package com.torvm.android.packet

/**
 * Utility object for assembling complete IPv4 packets with proper checksums.
 *
 * Builds TCP or UDP packets by:
 * 1. Creating the transport header
 * 2. Creating the IP header with correct total length
 * 3. Computing the transport-layer checksum (using pseudo-header)
 * 4. Computing the IP header checksum
 * 5. Concatenating all parts into a single byte array
 */
object PacketBuilder {

    private const val DEFAULT_TTL = 64
    private const val IP_HEADER_SIZE = 20

    /**
     * Build a complete IPv4/TCP packet.
     *
     * @param srcAddr source IPv4 address (4 bytes)
     * @param dstAddr destination IPv4 address (4 bytes)
     * @param srcPort source TCP port
     * @param dstPort destination TCP port
     * @param seqNum TCP sequence number (unsigned 32-bit stored as Long)
     * @param ackNum TCP acknowledgment number (unsigned 32-bit stored as Long)
     * @param flags TCP flags (e.g., TcpHeader.FLAG_SYN or TcpHeader.FLAG_ACK)
     * @param window TCP window size
     * @param payload TCP payload data
     * @param identification IP identification field
     * @return the complete IP packet as a byte array
     */
    fun buildTcpPacket(
        srcAddr: ByteArray,
        dstAddr: ByteArray,
        srcPort: Int,
        dstPort: Int,
        seqNum: Long,
        ackNum: Long,
        flags: Int,
        window: Int,
        payload: ByteArray = byteArrayOf(),
        identification: Int = 0
    ): ByteArray {
        val tcpHeaderSize = 20 // No options
        val totalLength = IP_HEADER_SIZE + tcpHeaderSize + payload.size

        // Build TCP header with checksum = 0 (will be computed below)
        val tcpHeader = TcpHeader(
            sourcePort = srcPort,
            destPort = dstPort,
            sequenceNumber = seqNum,
            ackNumber = ackNum,
            dataOffset = 5, // 20 bytes / 4
            reserved = 0,
            flags = flags,
            window = window,
            checksum = 0,
            urgentPointer = 0
        )

        // Build IP header with checksum = 0 (will be recalculated by toBytes())
        val ipHeader = IPv4Header(
            version = 4,
            ihl = 5,
            tos = 0,
            totalLength = totalLength,
            identification = identification,
            flags = 0x02, // Don't Fragment
            fragmentOffset = 0,
            ttl = DEFAULT_TTL,
            protocol = IPv4Header.PROTOCOL_TCP,
            headerChecksum = 0,
            sourceAddress = srcAddr.copyOf(),
            destAddress = dstAddr.copyOf()
        )

        // Compute TCP checksum using pseudo-header
        val tcpBytes = tcpHeader.toBytes()
        val tcpChecksum = Checksum.tcpUdpChecksum(ipHeader, tcpBytes, payload)

        // Update TCP header bytes with the computed checksum
        tcpBytes[16] = (tcpChecksum shr 8).toByte()
        tcpBytes[17] = tcpChecksum.toByte()

        // Serialize IP header (recalculates IP checksum internally)
        val ipBytes = ipHeader.toBytes()

        // Assemble the complete packet
        val packet = ByteArray(totalLength)
        System.arraycopy(ipBytes, 0, packet, 0, IP_HEADER_SIZE)
        System.arraycopy(tcpBytes, 0, packet, IP_HEADER_SIZE, tcpHeaderSize)
        if (payload.isNotEmpty()) {
            System.arraycopy(payload, 0, packet, IP_HEADER_SIZE + tcpHeaderSize, payload.size)
        }

        return packet
    }

    /**
     * Build a complete IPv4/UDP packet.
     *
     * @param srcAddr source IPv4 address (4 bytes)
     * @param dstAddr destination IPv4 address (4 bytes)
     * @param srcPort source UDP port
     * @param dstPort destination UDP port
     * @param payload UDP payload data
     * @return the complete IP packet as a byte array
     */
    fun buildUdpPacket(
        srcAddr: ByteArray,
        dstAddr: ByteArray,
        srcPort: Int,
        dstPort: Int,
        payload: ByteArray
    ): ByteArray {
        val udpLength = UdpHeader.HEADER_SIZE + payload.size
        val totalLength = IP_HEADER_SIZE + udpLength

        // Build UDP header with checksum = 0 (will be computed below)
        val udpHeader = UdpHeader(
            sourcePort = srcPort,
            destPort = dstPort,
            length = udpLength,
            checksum = 0
        )

        // Build IP header with checksum = 0 (will be recalculated by toBytes())
        val ipHeader = IPv4Header(
            version = 4,
            ihl = 5,
            tos = 0,
            totalLength = totalLength,
            identification = 0,
            flags = 0x02, // Don't Fragment
            fragmentOffset = 0,
            ttl = DEFAULT_TTL,
            protocol = IPv4Header.PROTOCOL_UDP,
            headerChecksum = 0,
            sourceAddress = srcAddr.copyOf(),
            destAddress = dstAddr.copyOf()
        )

        // Compute UDP checksum using pseudo-header
        val udpBytes = udpHeader.toBytes()
        val udpChecksum = Checksum.tcpUdpChecksum(ipHeader, udpBytes, payload)

        // Update UDP header bytes with the computed checksum
        udpBytes[6] = (udpChecksum shr 8).toByte()
        udpBytes[7] = udpChecksum.toByte()

        // Serialize IP header (recalculates IP checksum internally)
        val ipBytes = ipHeader.toBytes()

        // Assemble the complete packet
        val packet = ByteArray(totalLength)
        System.arraycopy(ipBytes, 0, packet, 0, IP_HEADER_SIZE)
        System.arraycopy(udpBytes, 0, packet, IP_HEADER_SIZE, UdpHeader.HEADER_SIZE)
        if (payload.isNotEmpty()) {
            System.arraycopy(payload, 0, packet, IP_HEADER_SIZE + UdpHeader.HEADER_SIZE, payload.size)
        }

        return packet
    }
}

File: ios/TorVMTunnel/Packet/PacketBuilder.swift
import Foundation

/// Utility for assembling complete IPv4 packets with proper checksums.
///
/// Builds TCP or UDP packets by:
/// 1. Creating the transport header
/// 2. Creating the IP header with correct total length
/// 3. Computing the transport-layer checksum (using pseudo-header)
/// 4. Computing the IP header checksum
/// 5. Concatenating all parts into a single Data
enum PacketBuilder {

    private static let defaultTTL: UInt8 = 64
    private static let ipHeaderSize = 20
    static let windowSize: UInt16 = 65535

    /// Build a complete IPv4/TCP packet.
    ///
    /// - Parameters:
    ///   - srcAddr: source IPv4 address (4 bytes)
    ///   - dstAddr: destination IPv4 address (4 bytes)
    ///   - srcPort: source TCP port
    ///   - dstPort: destination TCP port
    ///   - seqNum: TCP sequence number
    ///   - ackNum: TCP acknowledgment number
    ///   - flags: TCP flags (e.g., TcpHeader.flagSYN | TcpHeader.flagACK)
    ///   - window: TCP window size
    ///   - payload: TCP payload data
    ///   - identification: IP identification field
    /// - Returns: the complete IP packet as Data
    static func buildTcpPacket(
        srcAddr: Data,
        dstAddr: Data,
        srcPort: UInt16,
        dstPort: UInt16,
        seqNum: UInt32,
        ackNum: UInt32,
        flags: UInt16,
        window: UInt16,
        payload: Data = Data(),
        identification: UInt16 = 0
    ) -> Data {
        let tcpHeaderSize = 20 // No options
        let totalLength = ipHeaderSize + tcpHeaderSize + payload.count

        // Build TCP header with checksum = 0 (will be computed below)
        let tcpHeader = TcpHeader(
            sourcePort: srcPort,
            destPort: dstPort,
            sequenceNumber: seqNum,
            ackNumber: ackNum,
            dataOffset: 5, // 20 bytes / 4
            reserved: 0,
            flags: flags,
            window: window,
            checksum: 0,
            urgentPointer: 0
        )

        // Build IP header with checksum = 0 (will be recalculated by toBytes())
        let ipHeader = IPv4Header(
            version: 4,
            ihl: 5,
            tos: 0,
            totalLength: UInt16(totalLength),
            identification: identification,
            flags: 0x02, // Don't Fragment
            fragmentOffset: 0,
            ttl: defaultTTL,
            protocol: IPv4Header.protocolTCP,
            headerChecksum: 0,
            sourceAddress: Data(srcAddr),
            destAddress: Data(dstAddr)
        )

        // Compute TCP checksum using pseudo-header
        var tcpBytes = tcpHeader.toBytes()
        let tcpChecksum = Checksum.tcpUdpChecksum(ipHeader: ipHeader, transportHeader: tcpBytes, payload: payload)

        // Update TCP header bytes with the computed checksum
        tcpBytes[16] = UInt8(tcpChecksum >> 8)
        tcpBytes[17] = UInt8(tcpChecksum & 0xFF)

        // Serialize IP header (recalculates IP checksum internally)
        let ipBytes = ipHeader.toBytes()

        // Assemble the complete packet
        var packet = Data(count: totalLength)
        packet.replaceSubrange(0..<ipHeaderSize, with: ipBytes)
        packet.replaceSubrange(ipHeaderSize..<(ipHeaderSize + tcpHeaderSize), with: tcpBytes)
        if !payload.isEmpty {
            packet.replaceSubrange((ipHeaderSize + tcpHeaderSize)..<totalLength, with: payload)
        }

        return packet
    }

    /// Build a complete IPv4/UDP packet.
    ///
    /// - Parameters:
    ///   - srcAddr: source IPv4 address (4 bytes)
    ///   - dstAddr: destination IPv4 address (4 bytes)
    ///   - srcPort: source UDP port
    ///   - dstPort: destination UDP port
    ///   - payload: UDP payload data
    /// - Returns: the complete IP packet as Data
    static func buildUdpPacket(
        srcAddr: Data,
        dstAddr: Data,
        srcPort: UInt16,
        dstPort: UInt16,
        payload: Data
    ) -> Data {
        let udpLength = UdpHeader.headerSize + payload.count
        let totalLength = ipHeaderSize + udpLength

        // Build UDP header with checksum = 0 (will be computed below)
        let udpHeader = UdpHeader(
            sourcePort: srcPort,
            destPort: dstPort,
            length: UInt16(udpLength),
            checksum: 0
        )

        // Build IP header with checksum = 0 (will be recalculated by toBytes())
        let ipHeader = IPv4Header(
            version: 4,
            ihl: 5,
            tos: 0,
            totalLength: UInt16(totalLength),
            identification: 0,
            flags: 0x02, // Don't Fragment
            fragmentOffset: 0,
            ttl: defaultTTL,
            protocol: IPv4Header.protocolUDP,
            headerChecksum: 0,
            sourceAddress: Data(srcAddr),
            destAddress: Data(dstAddr)
        )

        // Compute UDP checksum using pseudo-header
        var udpBytes = udpHeader.toBytes()
        let udpChecksum = Checksum.tcpUdpChecksum(ipHeader: ipHeader, transportHeader: udpBytes, payload: payload)

        // Update UDP header bytes with the computed checksum
        udpBytes[6] = UInt8(udpChecksum >> 8)
        udpBytes[7] = UInt8(udpChecksum & 0xFF)

        // Serialize IP header (recalculates IP checksum internally)
        let ipBytes = ipHeader.toBytes()

        // Assemble the complete packet
        var packet = Data(count: totalLength)
        packet.replaceSubrange(0..<ipHeaderSize, with: ipBytes)
        packet.replaceSubrange(ipHeaderSize..<(ipHeaderSize + UdpHeader.headerSize), with: udpBytes)
        if !payload.isEmpty {
            packet.replaceSubrange((ipHeaderSize + UdpHeader.headerSize)..<totalLength, with: payload)
        }

        return packet
    }
}
